### PR TITLE
 Support Sidekiq Pro super_fetch 

### PR DESCRIPTION
**Pull request by mnovelo**
Distilling the discussions in [this issue](https://github.com/ixti/sidekiq-throttled/issues/111) and the work in [this fork](https://github.com/bookingexperts/sidekiq-throttled/tree/sidekiq-pro-support), this adds support for Sidekiq Pro's super_fetch for v1 of sidekiq-throttled and v7.x of Sidekiq Pro.

### CODE REVIEW

File: lib/sidekiq/throttled/patches/super_fetch.rb [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        end
+
+        # Returns list of non-paused queues to try to fetch jobs from.
+        #
+        # @note It may return an empty array.
+        # @return [Array<Array(String, String)>]
+        def active_queues
+          # Create a hash of throttled queues for fast lookup
+          throttled_queues = Throttled.cooldown&.queues&.to_a&.to_h { [_1, true] }
+          return super unless throttled_queues
```
Comment (ixti) [inline lib/sidekiq/throttled/patches/super_fetch.rb]: I think we can check if throttled queues is nil or empty here:
```suggestion
          return super if throttled_queues.nil? || throttled_queues.empty?
```

**Comment by ixti** [comment]
@mnovelo please add codecov path exclusion (`.github/codecov.yml`):

``` yaml
ignore:
  # Requires Sidekiq-Pro
  - lib/sidekiq/throttled/patches/super_fetch.rb
  - spec/lib/sidekiq/throttled/patches/super_fetch_spec.rb
```

File: spec/support/sidekiq.rb [hunk 2]
```diff
@@ -8,11 +8,81 @@
...
+rescue LoadError
+  # Sidekiq Pro is not available
+end
+
+$TESTING = true # rubocop:disable Style/GlobalVars
 
 REDIS_URL = ENV.fetch("REDIS_URL", "redis://localhost:6379")
 
-module JidGenerator
+module SidekiqHelper
```
Comment (ixti) [inline spec/support/sidekiq.rb]:
```suggestion
module SidekiqThrottledHelper
```

File: spec/support/sidekiq.rb [hunk 3]
```diff
@@ -54,15 +124,11 @@ def output
 end
 
 RSpec.configure do |config|
-  config.include JidGenerator
-  config.extend  JidGenerator
+  config.include SidekiqHelper
```
Comment (ixti) [inline spec/support/sidekiq.rb]:
```suggestion
  config.include SidekiqThrottledHelper
```

File: spec/lib/sidekiq/throttled/patches/super_fetch_spec.rb [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+begin
+  require "sidekiq/pro/super_fetch"
+  SIDEKIQ_PRO_AVAILABLE = true
+rescue LoadError
+  # Sidekiq Pro is not available
+  SIDEKIQ_PRO_AVAILABLE = false
+end
+
+RSpec.describe Sidekiq::Throttled::Patches::SuperFetch do
+  if SIDEKIQ_PRO_AVAILABLE
```
Comment (ixti) [inline spec/lib/sidekiq/throttled/patches/super_fetch_spec.rb]: As we're already loading sidekiq-pro in the patch - we can shorten this to:

```suggestion

RSpec.describe Sidekiq::Throttled::Patches::SuperFetch do
  if defined? Sidekiq::Pro::SuperFetch
```

Comment (mnovelo) [reply spec/lib/sidekiq/throttled/patches/super_fetch_spec.rb]: Really appreciate all the feedback! 💪🏽

File: spec/lib/sidekiq/throttled/patches/super_fetch_spec.rb [hunk 5]
```diff
@@ -0,0 +1,105 @@
+# frozen_string_literal: true
+
+require "sidekiq/throttled/patches/super_fetch"
+begin
+  require "sidekiq/pro/super_fetch"
+rescue LoadError
+  # Sidekiq Pro is not available
+end
```
Comment (ixti) [inline spec/lib/sidekiq/throttled/patches/super_fetch_spec.rb]: I don't think we need this:

```suggestion
```

After `require "sidekiq/throttled/patches/super_fetch"` we can safely assume we did that already

Comment (mnovelo) [reply spec/lib/sidekiq/throttled/patches/super_fetch_spec.rb]: You're right. Specs still pass without it 💪🏽

**Comment by ixti** [comment]
After taking another look, I think there's a better way to exclude Sidekiq-Pro specs when it's not available. Let's add this change to `spec_helper.rb`:

``` diff
diff --git a/spec/spec_helper.rb b/spec/spec_helper.rb
index cb34b71..661327d 100644
--- a/spec/spec_helper.rb
+++ b/spec/spec_helper.rb
@@ -61,6 +61,9 @@ RSpec.configure do |config|
   # metadata: `fit`, `fdescribe` and `fcontext`, respectively.
   config.filter_run_when_matching :focus
 
+  # Skip tests that require Sidekiq-Pro
+  config.filter_run_excluding sidekiq_pro: true unless defined? Sidekiq::Pro
+
   # Allows RSpec to persist some state between runs in order to support
   # the `--only-failures` and `--next-failure` CLI options. We recommend
   # you configure your source control system to ignore this file.
```

Then `spec/lib/sidekiq/throttled/patches/super_fetch_spec.rb` will become:

``` ruby
RSpec.describe Sidekiq::Throttled::Patches::SuperFetch, :sidekiq_pro do
  let(:base_queue) { "default" }
  ...
```

So that spec suite will run only when Sidekiq-Pro is available, or when one runs it explicitly with:

``` bash
bundle exec rspec --tag sidekiq_pro
```

**Review by ixti — APPROVED** [review]
Looks good to me. @freemanoid please take another look.

**Review by freemanoid — APPROVED** [review]

File: lib/sidekiq/throttled/patches/super_fetch.rb [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+          work.requeue
+        end
+
+        # Returns list of non-paused queues to try to fetch jobs from.
+        #
+        # @note It may return an empty array.
+        # @return [Array<Array(String, String)>]
+        def active_queues
+          # Create a hash of throttled queues for fast lookup
+          throttled_queues = Throttled.cooldown&.queues&.to_a&.to_h { [_1, true] }
```
Comment (ixti) [inline lib/sidekiq/throttled/patches/super_fetch.rb]: I know that it was me proposing to use `_1`, but I guess we can go with something that will be closer to Ruby-3.4 :D

```suggestion
          throttled_queues = Throttled.cooldown&.queues&.to_a&.to_h { |it| [it, true] }
```

Or actually decypher it:

```suggestion
          throttled_queues = Throttled.cooldown&.queues&.to_a&.to_h { |q| [q, true] }
```

**Comment by ixti** [comment]
@mnovelo I've added the above proposals (and a bit more) to a separate branch

**Comment by ixti** [comment]
Can you please check if the updated instructions make sense and specs pass for you https://github.com/ixti/sidekiq-throttled/tree/sidekiq-pro-super-fetch - and if so - I will merge it

**Comment by mnovelo** [comment]
Thanks! I was just wrapping them up and taking care of the Rubocop `Metrics/CyclomaticComplexity: Cyclomatic complexity for active_queues is too high.` warning. Will push in a moment

**Comment by ixti** [comment]
> Thanks! I was just wrapping them up and taking care of the Rubocop `Metrics/CyclomaticComplexity: Cyclomatic complexity for active_queues is too high.` warning. Will push in a moment

Oh. hm... I wonder why `Metrics/CyclomaticComplexity` fails. It's not that complex LOL

**Comment by mnovelo** [comment]
Just added a change that removes some redundant safe operators since `nil.to_a.to_h` becomes `{}`. That took care of it 😊

File: lib/sidekiq/throttled/patches/super_fetch.rb [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+          work.requeue
+        end
+
+        # Returns list of non-paused queues to try to fetch jobs from.
+        #
+        # @note It may return an empty array.
+        # @return [Array<Array(String, String)>]
+        def active_queues
+          # Create a hash of throttled queues for fast lookup
+          throttled_queues = Throttled.cooldown&.queues.to_a.to_h { |queue| [queue, true] }
```
Comment (ixti) [inline lib/sidekiq/throttled/patches/super_fetch.rb]: We don't need `#to_a` here. So it can be:

```suggestion
          throttled_queues = Throttled.cooldown&.queues&.to_h { |queue| [queue, true] }
```

Comment (ixti) [reply lib/sidekiq/throttled/patches/super_fetch.rb]: `Throttled.cooldown&.queues` is either `nil` or `Array`

Comment (mnovelo) [reply lib/sidekiq/throttled/patches/super_fetch.rb]: Nice! and `nil.to_h` is just `{}`

File: lib/sidekiq/throttled/patches/super_fetch.rb [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+          work.requeue
+        end
+
+        # Returns list of non-paused queues to try to fetch jobs from.
+        #
+        # @note It may return an empty array.
+        # @return [Array<Array(String, String)>]
+        def active_queues
+          # Create a hash of throttled queues for fast lookup
+          throttled_queues = Throttled.cooldown&.queues.to_h { |queue| [queue, true] }
```
Comment (ixti) [inline lib/sidekiq/throttled/patches/super_fetch.rb]: One last bit :D We can use `&.to_h` now without CyclomaticComplexity violation ;))

```suggestion
          throttled_queues = Throttled.cooldown&.queues&.to_h { |queue| [queue, true] }
```

Comment (mnovelo) [reply lib/sidekiq/throttled/patches/super_fetch.rb]: Removing the safe operator also removes the need to check for nil in the next line so that we can just call `throttled_queues.empty?`

Comment (ixti) [reply lib/sidekiq/throttled/patches/super_fetch.rb]: It does, indeed. But that means we will be creating empty Hash all the time when `cooldown` is `nil`

Comment (mnovelo) [reply lib/sidekiq/throttled/patches/super_fetch.rb]: oh good point!

**Comment by ixti** [comment]
I'm gonna squash merge this PR in a bit!

**Comment by mnovelo** [comment]
sorry, one moment. Specs aren't passing now

**Comment by ixti** [comment]
Oh.. Yeah. I'll wait.

**Comment by ixti** [comment]
> sorry, one moment. Specs aren't passing now

Make sure to `cp .rspec.sidekiq-pro .rspec-local`, by default `:sidekiq_pro` tests are disabled, and `.rspec.sidekiq-pro` is a version of `.rspec` that does not filter them out.

**Comment by mnovelo** [comment]
Oh it's the 6.5 specs that are failing. I had it set to skip Sidekiq Pro < 7 since it uses a different config. For some reason, now they're running

**Comment by mnovelo** [comment]
Weird fluke in my local environment. Now everything is back and passing 😖 Anyway, ready to merge!

**Comment by mnovelo** [comment]
@ixti I applied [this suggestion](https://github.com/ixti/sidekiq-throttled/pull/178/commits/15522030616bc16794c12ac0921d5b3f833a6b35) using the GitHub UI but it didn't get signed. Do you need me to fix that or does it not matter since you're gonna squash and merge?

**Comment by ixti** [comment]
> @ixti I applied [this suggestion](https://github.com/ixti/sidekiq-throttled/pull/178/commits/15522030616bc16794c12ac0921d5b3f833a6b35) using the GitHub UI but it didn't get signed. Do you need me to fix that or does it not matter since you're gonna squash and merge?

@mnovelo Yeah. Gonna squash merge it. So. don't worry about commits not being signed :D

**Comment by ixti** [comment]
@mnovelo Thank you very much! I'm gonna cut 1.3.0 release. and in 2.0.0 sidekiq 6.x will be fully dropped, so I am not worried about it :D

**Comment by ixti** [comment]
Released. Thank you very much to all who participated in this journey! :heart:
